common/metrics/metrics.go

```diff
@@ -0,0 +1,119 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"time"
+
+	"golang.org/x/exp/event"
+)
+
+// Mostly cribbed from
+// https://github.com/temporalio/sdk-go/blob/master/internal/common/metrics/handler.go
+// and adapted to depend on golang.org/x/exp/event
+type (
+	// MetricProvider represents the main dependency for instrumentation
+	MetricProvider interface {
+		// WithTags creates a new MetricProvder with provided []Tag
+		WithTags(...Tag) MetricProvider
+
+		// Counter obtains a counter for the given name.
+		Counter(string, ...MetricOption) CounterMetric
+
+		// Gauge obtains a gauge for the given name.
+		Gauge(string, ...MetricOption) GaugeMetric
+
+		// Timer obtains a timer for the given name.
+		Timer(string, ...MetricOption) TimerMetric
+
+		// Histogram obtains a histogram for the given name.
+		Histogram(string, ...MetricOption) HistogramMetric
+
+		// Tags returns registered []Tag
+		Tags() []Tag
+	}
+
+	// CounterMetric is an ever-increasing counter.
+	CounterMetric interface {
+		// Record increments the counter value.
+		Record(int64, ...Tag)
+	}
+	// GaugeMetric can be set to any float.
+	GaugeMetric interface {
+		// Record updates the gauge value.
+		Record(float64, ...Tag)
+	}
+
+	// TimerMetric records time durations.
+	TimerMetric interface {
+		// Record sets the timer value.
+		Record(time.Duration, ...Tag)
+	}
+
+	// HistogramMetric records a distribution of values.
+	HistogramMetric interface {
+		// Record adds a value to the distribution
+		Record(int64, ...Tag)
+	}
+
+	CounterMetricFunc   func(int64, ...Tag)
+	GaugeMetricFunc     func(float64, ...Tag)
+	TimerMetricFunc     func(time.Duration, ...Tag)
+	HistogramMetricFunc func(int64, ...Tag)
+
+	MetricOption interface {
+		apply(*event.MetricOptions)
+	}
+
+	MetricOptionFunc func(*event.MetricOptions)
+)
+
+// WithMetricUnit provides MetricUnit MetricOption for metric
+func WithMetricUnit(unit MetricUnit) MetricOption {
+	return MetricOptionFunc(func(mo *event.MetricOptions) {
+		mo.Unit = event.Unit(unit)
```

Comment:
const values of our MetricUnit is different from event.Unit.

Reply:
This is still in draft while I work through those last few details.

The types are both typedefs off string, but what is needed is new constants mapping to the new values for this dependency.

Reply:
I'm not sure if we persist those value anywhere. If not, we can just change our const value to match event.Unit, or just type def our MetricUnit as event.Unit.

Reply:
Will do, the event library and Otel constants for these dimension units also match.
Makes sense we can just reconcile to the same values when the older interface is fully deprecated.